File: python/src/dexcost/idempotency.py
"""Caller-controlled, privacy-safe idempotency for durable event capture."""

from __future__ import annotations

import contextvars
import hashlib
import json
import threading
import uuid
from collections.abc import Generator
from contextlib import contextmanager
from dataclasses import dataclass, field
from typing import Any

from dexcost.models.event import Event

_EVENT_NAMESPACE = uuid.UUID("ee9858ce-fc4e-5c97-a803-2ea9df316d5c")


@dataclass
class _IdempotencyScope:
    key: str
    next_occurrence: int = 0
    lock: threading.Lock = field(default_factory=threading.Lock)

    def capture(self) -> CapturedIdempotencyKey:
        with self.lock:
            occurrence = self.next_occurrence
            self.next_occurrence += 1
        return CapturedIdempotencyKey(self.key, occurrence)


@dataclass(frozen=True)
class CapturedIdempotencyKey:
    """One stable occurrence reserved from an ambient caller-key scope."""

    key: str
    occurrence: int


IdempotencyKey = str | CapturedIdempotencyKey


_KEY_CONTEXT: contextvars.ContextVar[_IdempotencyScope | None] = contextvars.ContextVar(
    "dexcost_idempotency_key", default=None
)
_HASH_DETAIL = "_dexcost_idempotency_sha256"
_OCCURRENCE_DETAIL = "_dexcost_idempotency_occurrence"


def _validate_key(key: str) -> str:
    if not isinstance(key, str):
        raise TypeError("idempotency key must be a string")
    if not 1 <= len(key) <= 255:
        raise ValueError("idempotency key must contain 1 to 255 characters")
    if any(ord(character) < 0x21 or ord(character) > 0x7E for character in key):
        raise ValueError("idempotency key must contain visible ASCII characters only")
    return key


def get_idempotency_key() -> str | None:
    """Return the caller key active in this context, if any."""
    scope = _KEY_CONTEXT.get()
    return None if scope is None else scope.key


def capture_idempotency_key() -> CapturedIdempotencyKey | None:
    """Reserve one deterministic operation occurrence from the active scope."""
    scope = _KEY_CONTEXT.get()
    return None if scope is None else scope.capture()


def set_idempotency_key(
    key: str | None,
) -> contextvars.Token[Any]:
    """Set or clear the caller key and return a token for precise restoration."""
    return _KEY_CONTEXT.set(None if key is None else _IdempotencyScope(_validate_key(key)))


@contextmanager
def idempotency_key(key: str) -> Generator[None, None, None]:
    """Scope one stable caller key to captured operations."""
    token = set_idempotency_key(key)
    try:
        yield
    finally:
        _KEY_CONTEXT.reset(token)


def _event_identity(event: Event) -> str:
    details = event.details
    identity: dict[str, Any] = {
        "event_type": event.event_type,
        "provider": event.provider,
        "model": event.model,
        "service_name": event.service_name,
        "operation_name": details.get("attribution_operation_name"),
        "resource_type": details.get("attribution_resource_type"),
        "resource_id": details.get("attribution_resource_id"),
        "capability": details.get("attribution_capability"),
    }
    return json.dumps(identity, sort_keys=True, separators=(",", ":"))


def apply_event_idempotency(event: Event, key: IdempotencyKey | None = None) -> Event:
    """Stamp a deterministic opaque event ID from the active caller key."""
    # Storage backends defensively apply the policy again. Preserve a stamp
    # already reserved by the operation wrapper instead of consuming another
    # occurrence from the ambient scope.
    if idempotency_hash(event) is not None:
        return event
    captured = capture_idempotency_key() if key is None else key
    if captured is None:
        return event
    if isinstance(captured, CapturedIdempotencyKey):
        resolved_key = _validate_key(captured.key)
        occurrence: int | None = captured.occurrence
    else:
        resolved_key = _validate_key(captured)
        occurrence = None
    key_hash = hashlib.sha256(resolved_key.encode("ascii")).hexdigest()
    original_event_id = str(event.event_id)
    identity_parts = [str(event.task_id), key_hash, _event_identity(event)]
    if occurrence is not None:
        identity_parts.insert(2, str(occurrence))
    event.event_id = uuid.uuid5(
        _EVENT_NAMESPACE,
        "\0".join(identity_parts),
    )
    event.details = {
        **event.details,
        _HASH_DETAIL: key_hash,
        **({_OCCURRENCE_DETAIL: occurrence} if occurrence is not None else {}),
    }
    for identity_key in ("attribution_operation_id", "attribution_attempt_id"):
        if event.details.get(identity_key) == original_event_id:
            event.details[identity_key] = str(event.event_id)
    return event


def idempotency_hash(event: Event) -> str | None:
    """Return the opaque key hash stamped on an event, if valid."""
    value = event.details.get(_HASH_DETAIL)
    if (
        isinstance(value, str)
        and len(value) == 64
        and all(character in "0123456789abcdef" for character in value)
    ):
        return value
    return None


def equivalent_idempotent_event(left: Event, right: Event) -> bool:
    """Compare repeated capture bodies while ignoring their new wall timestamp."""
    if idempotency_hash(left) is None or idempotency_hash(left) != idempotency_hash(right):
        return False
    left_dict = left.to_dict()
    right_dict = right.to_dict()
    left_dict.pop("occurred_at", None)
    right_dict.pop("occurred_at", None)
    return bool(left_dict == right_dict)


__all__ = [
    "CapturedIdempotencyKey",
    "IdempotencyKey",
    "apply_event_idempotency",
    "capture_idempotency_key",
    "equivalent_idempotent_event",
    "get_idempotency_key",
    "idempotency_hash",
    "idempotency_key",
    "set_idempotency_key",
]

File: typescript/src/core/idempotency.ts
import { AsyncLocalStorage } from "node:async_hooks";
import { createHash } from "node:crypto";
import { eventToDict, type CostEvent } from "./models.js";

interface IdempotencyScope {
  readonly key: string;
  nextOccurrence: number;
}

interface IdempotencyContext {
  readonly scope?: IdempotencyScope;
}

export interface CapturedIdempotencyKey {
  readonly key: string;
  readonly occurrence: number;
}

export type IdempotencyKey = string | CapturedIdempotencyKey;

/**
 * One-shot token returned by {@link setIdempotencyKey}.
 *
 * Call `reset()` to restore the exact previous scope, including its next
 * occurrence counter. Reset the token before yielding to another async
 * continuation. For scopes that cross `await`, use
 * {@link runWithIdempotencyKey}; Node's callback-scoped AsyncLocalStorage API
 * is what preserves already-created child contexts while restoring callers.
 */
export interface IdempotencyKeyToken {
  readonly previousKey: string | undefined;
  reset(): void;
}

interface IdempotencyTokenState {
  readonly previous: IdempotencyContext | undefined;
  readonly installed: IdempotencyContext;
  used: boolean;
}

const keyStore = new AsyncLocalStorage<IdempotencyContext | undefined>();
const tokenStates = new WeakMap<object, IdempotencyTokenState>();
const EVENT_NAMESPACE = "ee9858ce-fc4e-5c97-a803-2ea9df316d5c";

function validateKey(key: string): string {
  if (typeof key !== "string" || key.length < 1 || key.length > 255) {
    throw new Error("idempotency key must contain 1 to 255 characters");
  }
  for (const character of key) {
    const code = character.charCodeAt(0);
    if (code < 0x21 || code > 0x7e) {
      throw new Error("idempotency key must contain visible ASCII characters only");
    }
  }
  return key;
}

function activeContext(): IdempotencyContext | undefined {
  return keyStore.getStore();
}

export function getIdempotencyKey(): string | undefined {
  return activeContext()?.scope?.key;
}

/** Reserve one deterministic operation occurrence from the active scope. */
export function captureIdempotencyKey(): CapturedIdempotencyKey | undefined {
  const scope = activeContext()?.scope;
  if (scope === undefined) return undefined;
  const occurrence = scope.nextOccurrence;
  scope.nextOccurrence += 1;
  return { key: scope.key, occurrence };
}

function createToken(
  previous: IdempotencyContext | undefined,
  installed: IdempotencyContext,
): IdempotencyKeyToken {
  const token: IdempotencyKeyToken = {
    previousKey: previous?.scope?.key,
    reset(): void {
      restoreToken(token);
    },
  };
  tokenStates.set(token, { previous, installed, used: false });
  return Object.freeze(token);
}

function restoreToken(token: IdempotencyKeyToken): void {
  const state = tokenStates.get(token);
  if (state === undefined) throw new Error("invalid idempotency key token");
  if (state.used) throw new Error("idempotency key token has already been reset");
  if (activeContext() !== state.installed) {
    throw new Error("idempotency key tokens must be reset in LIFO order in their originating async context");
  }
  state.used = true;
  // Restore only this execution context. Async resources created while the
  // installed context was active have already captured it and must retain it.
  keyStore.enterWith(state.previous);
}

/**
 * Set or clear the caller key for the remaining async context.
 *
 * The returned token restores the exact prior scope and occurrence counter.
 * Tokens are single-use, synchronous, and must be restored in LIFO order.
 * Use {@link runWithIdempotencyKey} for a scope that crosses `await`.
 */
export function setIdempotencyKey(key?: string): IdempotencyKeyToken {
  const previous = activeContext();
  const installed: IdempotencyContext = {
    scope: key === undefined
      ? undefined
      : { key: validateKey(key), nextOccurrence: 0 },
  };
  keyStore.enterWith(installed);
  return createToken(previous, installed);
}

export function runWithIdempotencyKey<T>(key: string, fn: () => T): T {
  return keyStore.run({
    scope: { key: validateKey(key), nextOccurrence: 0 },
  }, fn);
}

export function idempotencyHash(key: string): string;
export function idempotencyHash(event: CostEvent): string | undefined;
export function idempotencyHash(value: string | CostEvent): string | undefined {
  if (typeof value === "string") {
    return createHash("sha256").update(validateKey(value), "ascii").digest("hex");
  }
  const hash = value.details["_dexcost_idempotency_sha256"];
  return typeof hash === "string" && /^[0-9a-f]{64}$/.test(hash) ? hash : undefined;
}

function uuidBytes(value: string): Buffer {
  return Buffer.from(value.replaceAll("-", ""), "hex");
}

function uuid5(namespace: string, input: string): string {
  const bytes = createHash("sha1").update(uuidBytes(namespace)).update(input, "utf8").digest().subarray(0, 16);
  bytes[6] = ((bytes[6] ?? 0) & 0x0f) | 0x50;
  bytes[8] = ((bytes[8] ?? 0) & 0x3f) | 0x80;
  const raw = bytes.toString("hex");
  return `${raw.slice(0, 8)}-${raw.slice(8, 12)}-${raw.slice(12, 16)}-${raw.slice(16, 20)}-${raw.slice(20)}`;
}

function stableJson(value: unknown): string {
  if (Array.isArray(value)) return `[${value.map(stableJson).join(",")}]`;
  if (value !== null && typeof value === "object") {
    return `{${Object.entries(value as Record<string, unknown>)
      .sort(([left], [right]) => left.localeCompare(right))
      .map(([key, item]) => `${JSON.stringify(key)}:${stableJson(item)}`).join(",")}}`;
  }
  return JSON.stringify(value);
}

export function applyEventIdempotency(
  event: CostEvent,
  key: IdempotencyKey | undefined = captureIdempotencyKey(),
): CostEvent {
  // The operation wrapper may already have reserved and stamped this event.
  // Defensive storage/capability stamping must not consume another occurrence.
  if (idempotencyHash(event) !== undefined || key === undefined) return event;
  const resolvedKey = typeof key === "string" ? key : key.key;
  const occurrence = typeof key === "string" ? undefined : key.occurrence;
  const hash = idempotencyHash(resolvedKey);
  const identity = stableJson({
    event_type: event.eventType,
    provider: event.provider ?? null,
    model: event.model ?? null,
    service_name: event.serviceName ?? null,
    operation_name: event.details["attribution_operation_name"] ?? null,
    resource_type: event.details["attribution_resource_type"] ?? null,
    resource_id: event.details["attribution_resource_id"] ?? null,
    capability: event.details["attribution_capability"] ?? null,
  });
  const oldId = event.eventId;
  const occurrencePart = occurrence === undefined ? "" : `${occurrence}\0`;
  event.eventId = uuid5(EVENT_NAMESPACE, `${event.taskId}\0${hash}\0${occurrencePart}${identity}`);
  event.details = {
    ...event.details,
    _dexcost_idempotency_sha256: hash,
    ...(occurrence === undefined ? {} : { _dexcost_idempotency_occurrence: occurrence }),
  };
  for (const field of ["attribution_operation_id", "attribution_attempt_id"]) {
    if (event.details[field] === oldId) event.details[field] = event.eventId;
  }
  return event;
}

export function equivalentIdempotentEvent(left: CostEvent, right: CostEvent): boolean {
  const leftHash = idempotencyHash(left);
  if (leftHash === undefined || leftHash !== idempotencyHash(right)) return false;
  const leftValue = eventToDict(left);
  const rightValue = eventToDict(right);
  delete leftValue["occurred_at"];
  delete rightValue["occurred_at"];
  return stableJson(leftValue) === stableJson(rightValue);
}
